src/hazelweb/UndoHistory.re

```diff
@@ -0,0 +1,27 @@
+module ZList = GeneralUtil.ZList;
+
+type t = ZList.t(Statics.edit_state, Statics.edit_state);
+
+let add_history = (undo_history: t, edit_state: Statics.edit_state): t => {
+  /* first add new edit state to the end, then shift_next */
+  let add_new = (
+    ZList.prj_prefix(undo_history),
+    ZList.prj_z(undo_history),
+    [edit_state],
+  );
+  ZList.shift_next(add_new);
```

Comment:
change shift_next and shift_prev to have an option type, and then do the necessary refactorings